File: src/test/java/simplifying/conditional/expressions/replaceConditionalWithPolymorphism.java
package simplifying.conditional.expressions;

import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;

public class replaceConditionalWithPolymorphism {
    @Test
    public void calculator_should_support_add() {
        int result = Calculator.calculate(1, 2, Calculator.ADD);
        assertThat(result).isEqualTo(3);
    }

    @Test
    public void calculator_should_support_multiply() {
        int result = Calculator.calculate(90, 4, Calculator.MULTIPLY);
        assertThat(result).isEqualTo(360);
    }

    @Test
    public void calculator_should_support_divide() {
        int result = Calculator.calculate(84, 4, Calculator.DIVIDE);
        assertThat(result).isEqualTo(21);
    }

    @Test
    public void calculator_should_support_subtract() {
        int result = Calculator.calculate(84, 23, Calculator.SUBTRACT);
        assertThat(result).isEqualTo(61);
    }

    @Test
    public void calculator_should_throw_exception_on_unsupported_operator() {
        assertThatExceptionOfType(IllegalArgumentException.class)
                .isThrownBy(() -> Calculator.calculate(84, 23, "unsupported exception"));
    }
}

File: src/scala/src/test/scala/simplifying/conditional/expressions/replaceConditionalWithPolymorphism.scala
package simplifying.conditional.expressions

import org.scalatest.TryValues.convertTryToSuccessOrFailure
import org.scalatest.funsuite.AnyFunSuite

class replaceConditionalWithPolymorphism extends AnyFunSuite {
  test("calculator should support add") {
    assert(Calculator.calculate(1, 2, operator = "add").success.value == 3)
  }

  test("calculator should support multiply") {
    assert(
      Calculator.calculate(90, 4, operator = "multiply").success.value == 360
    )
  }

  test("calculator should support divide") {
    assert(Calculator.calculate(84, 4, operator = "divide").success.value == 21)
  }

  test("calculator should support subtract") {
    assert(
      Calculator.calculate(84, 23, operator = "subtract").success.value == 61
    )
  }

  test("calculator should throw exception on unsupported operator") {
    assert(
      Calculator
        .calculate(84, 23, "unsupported operator")
        .failure
        .exception
        .getMessage == "Not supported operator"
    )
  }
}
